README.md
# gogo-shell

> A Node wrapper for sending Gogo commands to an OSGi environment


## Install

```
$ npm install --save gogo-shell
```


## Usage

Note: a Gogo shell must be running before invoking the connect method, otherwise a connection error will be thrown.

```js
var GogoShell = require('gogo-shell');

var gogoShell = new GogoShell();

var config = {
    port: 11311
};

gogoShell.connect(config)
    .then(function() {
        return gogoShell.sendCommand('lb');
    })
    .then(function(data) {
        // Do something with response data

        gogoShell.end();
    });
```


## API

GogoShell is an instance of [net.Socket](https://nodejs.org/api/net.html#net_class_net_socket) and inherits all it's properties, methods, and events.

### new GogoShell([options])

#### options

Type: `object`

The `GogoShell` constructor can also be passed the same configuration options as the [net.Socket constructor](https://nodejs.org/api/net.html#net_new_net_socket_options).

##### debug

type: `boolean`<br>

If set to true, prints all commands sent to the server, and all response data from the server to stdout.

#### examples

```js
var GogoShell = require('gogo-shell');

var gogoShell = new GogoShell({
    debug: true
});
```

### connect([options])

Connects to a TCP server which allows you to send Gogo commands to an OSGi environment. Returns a Promise which resolves when the server is ready to receive commands.

#### options

Type: `object`

##### host

Type: `string`<br>
Default: `127.0.0.1`

##### port

Type: `number`<br>
Required: `true`

See [socket.connect](https://nodejs.org/api/net.html#net_socket_connect_options_connectlistener) method for other options and their default values.

#### examples

```js
var GogoShell = require('gogo-shell');

var gogoShell = new GogoShell();

var config = {
    port: 11311
};

gogoShell.connect(config)
    .then(function() {
        // gogoShell.sendCommand(...);
    });
```


### help([command])

Returns Promise that resolves with array of available commands, or object literal containing information on specified command.

##### command

Type: `string`

The Gogo command to retieve help info for.

#### examples

```js
gogoShell.help()
    .then(function(data) {
        // data = array of available commands
    });
```

```js
gogoShell.help('install')
    .then(function(data) {
        // data = object literal containing api information for specified command
    });
```


### sendCommand(command, [options])

Sends Gogo command to an OSGi environment. Returns a Promise that resolves with the response data.

##### command

Type: `string`

The Gogo command that will be sent to the OSGi environment.

##### options

Type: `string`

Additional parameters, flags, and options that will be joined with the command argument.

#### examples

```js
gogoShell.sendCommand('lb', '-s')
    .then(function(data) {
        // data = list of installed bundles with symbolic name
    });
```

Chaining commands.

```js
var GogoShell = require('gogo-shell');

var gogoShell = new GogoShell();

var config = {
    port: 11311
};

gogoShell.connect(config)
    .then(function() {
        return gogoShell.sendCommand('lb');
    })
    .then(function(data) {
        // data = response from 'lb'

        return gogoShell.sendCommand('another command');
    })
    .then(function(data) {
        // data = response from 'another command'

        // Using socket.end for closing connection, otherwise Node process wouldn't end
        gogoShell.end();
    });
```


MIT


CHANGELOG.md
## Next Release
* No changes
***

## v0.0.4 - March 9, 2016
* New: added debug option to constructor for printing all commands and response data to stdout

## v0.0.3 - March 7, 2016
* Updated package.json

## v0.0.2 - March 7, 2016
* Optimized code example formatting for npm page

## v0.0.1 - March 7, 2016
* Initial publish


test/index.js
'use strict';

var _ = require('lodash');
var chai = require('chai');
var GogoShell = require('../index.js');
var net = require('net');
var path = require('path');

var assert = chai.assert;

describe('GogoShell', function() {
	var gogoServer;
	var gogoShell;

	describe('connect', function() {
		it('should return connection error', function(done) {
			startGogo({
				port: 1234
			});

			gogoShell.on('error', function(err) {
				assert.equal(err.code, 'ECONNREFUSED');
				assert.equal(err.address, '127.0.0.1');
				assert.equal(err.port, 1234);

				stopGogo(done);
			});
		});

		it('should fire ready event', function(done) {
			startGogo({
				port: 1337
			}, true);

			gogoShell.on('ready', function() {
				stopGogo(done);
			});
		});

		it('should return promise that resolves when connection is ready', function(done) {
			startGogo({
				port: 1337
			}, true)
				.then(function() {
					return gogoShell.sendCommand('command');
				})
				.then(function(data) {
					stopGogo(done);
				});
		});
	});

	describe('help', function() {
		it('should return array of available gogo commands', function(done) {
			startGogo({
				port: 1337
			}, true)
				.then(function() {
					return gogoShell.help();
				})
				.then(function(data) {
					assert.deepEqual(data, ['gogo:echo', 'gogo:format', 'gogo:getopt']);

					stopGogo(done);
				});
		});

		it('should parse and return info on specific command', function(done) {
			startGogo({
				port: 1337
			}, true)
				.then(function() {
					return gogoShell.help('command');
				})
				.then(function(data) {
					assert(!_.isUndefined(data.raw));
					assert.equal(data.command, 'command');
					assert.equal(data.description, 'command - command that does something');
					assert.equal(data.flags.length, 2);
					assert.equal(data.parameters.length, 1);
					assert.equal(data.scope, 'obr');

					stopGogo(done);
				});
		});
	});

	describe('sendCommand', function() {
		it('should return correct data from server as Promise', function(done) {
			startGogo({
				port: 1337
			}, true);

			gogoShell.on('ready', function() {
				gogoShell.sendCommand('command')
					.then(function(data) {
						assertDefaultData(data);

						stopGogo(done);
					});
			});
		});

		it('should catch error for sending too many commands at once', function(done) {
			startGogo({
				port: 1337
			}, true);

			gogoShell.on('ready', function() {
				gogoShell.sendCommand('async')
					.then(function(data) {
						assertAsyncData(data);

						stopGogo(done);
					});

				gogoShell.sendCommand('command')
					.catch(function(err) {
						assert.equal(err.message, 'Only one command can be sent at a time. "async" hasn\'t finished.');
					});
			});
		});

		it('should be able to chain commands', function(done) {
			startGogo({
				port: 1337
			}, true)
				.then(function() {
					return gogoShell.sendCommand('async');
				})
				.then(function(data) {
					assertAsyncData(data);

					return gogoShell.sendCommand('command');
				})
				.then(function(data) {
					assertDefaultData(data);

					return gogoShell.sendCommand('async');
				})
				.then(function(data) {
					assertAsyncData(data);

					stopGogo(done);
				});
		});

		it('should join arguments to create command string', function(done) {
			startGogo({
				port: 1337
			}, true)
				.then(function() {
					return gogoShell.sendCommand('joined', 'arguments', 'command', '--test');
				})
				.then(function(data) {
					assert.equal(data, 'joined arguments data\ng! ');

					done();
				});
		})
	});

	var helpCommandData = 'command - command that does something\n' +
		'  scope: obr\n' +
		'  flags:\n' +
		'    -f, --flag   does something\n' +
		'    -o, --otherflag   does something else\n' +
		'  parameters:\n' +
		'    String[]   ( <this> | <that> | <the-other> )[@<version>] ...\ng! ';

	function assertAsyncData(data) {
		assert.equal(data, 'chunk 1\nchunk 2\ndata\ng! ');
	}

	function assertDefaultData(data) {
		assert.equal(data, 'data\ng! ');
	}

	function startGogo(config, startServer) {
		if (startServer) {
			var defaultResponseData = 'data\ng! ';

			var handshakeBuffer = new Buffer([255, 251, 24, 255, 250, 24, 0, 86, 84, 50, 50, 48, 255, 240]);

			gogoServer = net.createServer(function(socket) {
				socket.on('data', function(data) {
					if (_.isEqual(data, handshakeBuffer)) {
						socket.write('Welcome to Apache Felix Gogo\n\ng! ');
					}
					else if (data.indexOf('async') > -1) {
						socket.write('chunk 1\n');

						setTimeout(function() {
							socket.write('chunk 2\n');
						}, 150);

						setTimeout(function() {
							socket.write(defaultResponseData);
						}, 300);
					}
					else if (data.indexOf('help command') > -1) {
						socket.write(helpCommandData);
					}
					else if (data.indexOf('help') > -1) {
						socket.write('gogo:echo\n' +
							'gogo:format\n' +
							'gogo:getopt\n' +
							'g! ');
					}
					else if (data.indexOf('joined arguments command --test') > -1) {
						socket.write('joined arguments data\ng! ');
					}
					else {
						socket.write(defaultResponseData);
					}
				});
			});

			gogoServer.listen({
				host: '127.0.0.1',
				port: 1337
			});
		}

		gogoShell = new GogoShell();

		return gogoShell.connect(config);
	}

	function stopGogo(done) {
		if (gogoShell) {
			gogoShell.end();
		}

		if (gogoServer) {
			gogoServer.close(done);
		}
		else {
			done();
		}

		gogoServer = null;
		gogoShell = null;
	}
});


test/integration.js
'use strict';

var _ = require('lodash');
var chai = require('chai');
var GogoShell = require('../index.js');
var net = require('net');
var path = require('path');

var assert = chai.assert;

describe('GogoShell - Liferay integeration', function() {
	var command;
	var gogoServer;
	var gogoShell;

	after(function() {
		stopGogo();
	});

	it('should attempt to connect with localhost:11311', function(done) {
		startGogo({
			port: 11311
		})
		.then(function() {
			done();
		});

		gogoShell.on('error', function(err) {
			assert.equal(err.code, 'ECONNREFUSED');

			done();
		});
	});

	it('should return list of installed packages', function(done) {
		if (!checkConnection()) return done();

		gogoShell.sendCommand('lb -s')
			.then(function(data) {
				assert(data);

				done();
			});
	});

	it('should return array of available gogo commands', function(done) {
		if (!checkConnection()) return done();

		gogoShell.help()
			.then(function(data) {
				assert(_.isArray(data));

				command = data[1];

				done();
			});
	});

	it('should parse and return info on specific command', function(done) {
		if (!checkConnection()) return done();

		gogoShell.help(command)
			.then(function(data) {
				assert(_.isObject(data));

				assert(data.description);
				assert(data.raw);
				assert(data.scope);

				done();
			});
	});

	function checkConnection() {
		var connected = true;

		if (!gogoShell.ready) {
			connected = false;

			console.log('    SKIPPING TEST');
		}

		return connected;
	}

	function startGogo(config) {
		gogoShell = new GogoShell();

		return gogoShell.connect(config);
	}

	function stopGogo(done) {
		if (gogoShell) {
			gogoShell.end();
		}

		gogoShell = null;
	}
});


index.js
'use strict';

var _ = require('lodash');
var EventEmitter = require('events').EventEmitter;
var net = require('net');
var path = require('path');
var Promise = require('bluebird');

var Socket = net.Socket;

var DEFAULT_CONNECT_CONFIG = {
	host: '127.0.0.1'
};

var STR_DATA = 'data';

var STR_NEWLINE = '\n';

var GogoShell = function(config) {
	config = config || {};

	Socket.call(this, config);

	this.debug = config.debug || false;

	this.init();
};

GogoShell.prototype = _.create(Socket.prototype, {
	init: function() {
		this.ready = false;

		this.on('connect', this._onConnect.bind(this))
	},

	connect: function(config) {
		var instance = this;

		config = _.assign({}, DEFAULT_CONNECT_CONFIG, config);

		Socket.prototype.connect.call(this, config);

		return new Promise(function(resolve, reject) {
			instance.once('ready', resolve);
		});
	},

	help: function(command) {
		var parser = command ? this._parseHelpCommandData : this._parseHelpData;

		return this.sendCommand(command ? 'help ' + command : 'help')
			.then(function(data) {
				return parser(data, command);
			});
	},

	sendCommand: function(command) {
		var instance = this;

		if (arguments.length > 1) {
			command = _.join(arguments, ' ');
		}

		return new Promise(function(resolve, reject) {
			if (instance.active) {
				reject(new Error('Only one command can be sent at a time. "' + instance.currentCommand + '" hasn\'t finished.'));
			}

			instance.active = true;
			instance.currentCommand = command;

			instance.on(STR_DATA, instance._getCommandDataListener(resolve));

			command = _.endsWith(command, STR_NEWLINE) ? command : command + STR_NEWLINE;

			instance.write(command);
		});
	},

	_getCommandDataListener: function(resolve) {
		var instance = this;

		var dataBuffer = [];

		var dataListener = function(data) {
			data = data.toString();

			dataBuffer.push(data);

			if (data.match(/g\!/)) {
				instance.removeListener(STR_DATA, dataListener);

				instance.active = false;
				instance.currentCommand = null;

				resolve(dataBuffer.join(''));
			}
		};

		return dataListener;
	},

	_onConnect: function() {
		var instance = this;

		var dataListener = function(data) {
			data = data.toString();

			if (!instance.ready && data.indexOf('g!') > -1) {
				instance.ready = true;

				instance.removeListener(STR_DATA, dataListener);

				instance.emit('ready');
			}
		};

		this.on(STR_DATA, dataListener);

		if (this.debug) {
			this.on(STR_DATA, function(data) {
				process.stdout.write(data.toString());
			});
		}

		this.write(new Buffer([255, 251, 24]));

		this.write(new Buffer([255, 250, 24, 0, 86, 84, 50, 50, 48, 255, 240]));
	},

	_parseHelpCommandData: function(data, command) {
		var currentGroup;
		var groupRegex = /(flags|options|parameters|scope):\s*(\S*)/;

		return _.reduce(data.split(STR_NEWLINE), function(result, line, index) {
			if (index == 0) {
				result.description = line;
				result.raw = data;
			}
			else if (groupRegex.test(line)) {
				var match = line.match(groupRegex);

				currentGroup = match[1];

				if (match[2]) {
					result[currentGroup] = match[2];
				}
				else {
					result[currentGroup] = [];
				}
			}
			else if (currentGroup && line.indexOf('g!') < 0) {
				result[currentGroup].push(_.trim(line));
			}

			return result;
		}, {
			command: command
		});
	},

	_parseHelpData: function(data) {
		return _.reduce(data.split(STR_NEWLINE), function(commands, line, index) {
			if (line.indexOf('g!') < 0) {
				commands.push(_.trim(line));
			}

			return commands;
		}, []);
	}
});

module.exports = GogoShell;
